File: packages/server/injector/rust/src/ast.rs
/* packages/server/injector/rust/src/ast.rs */

#[derive(Debug)]
pub(crate) enum AstNode {
  Text(String),
  Slot { path: String, mode: SlotMode },
  Attr { path: String, attr_name: String },
  StyleProp { path: String, css_property: String },
  If { path: String, then_nodes: Vec<AstNode>, else_nodes: Vec<AstNode> },
  Each { path: String, body_nodes: Vec<AstNode> },
  Match { path: String, branches: Vec<(String, Vec<AstNode>)> },
}

#[derive(Debug)]
pub(crate) enum SlotMode {
  Text,
  Html,
}

File: src/server/injector/native/src/ast.ts
/* src/server/injector/native/src/ast.ts */

export type AstNode =
  | TextNode
  | SlotNode
  | AttrNode
  | StylePropNode
  | IfNode
  | EachNode
  | MatchNode;

export interface TextNode {
  type: "text";
  value: string;
}

export interface SlotNode {
  type: "slot";
  path: string;
  mode: "text" | "html";
}

export interface AttrNode {
  type: "attr";
  path: string;
  attrName: string;
}

export interface StylePropNode {
  type: "styleProp";
  path: string;
  cssProperty: string;
}

export interface IfNode {
  type: "if";
  path: string;
  thenNodes: AstNode[];
  elseNodes: AstNode[];
}

export interface EachNode {
  type: "each";
  path: string;
  bodyNodes: AstNode[];
}

export interface MatchNode {
  type: "match";
  path: string;
  branches: Map<string, AstNode[]>;
}
